### PR TITLE
Add tests and new rules for machine learning libraries 

### PR DESCRIPTION
**Pull request by suhacker1**
I added the following rules for machine learning libraries such as PyTorch, NumPy, and Pandas. 

Rule ID | Language | What it Finds
--- | --- | ---
[automatic-memory-pinning](python/automatic-memory-pinning.yml) | Python | Memory is not automatically pinned
[lxml-in-pandas](python/lxml-in-pandas.yml) | Python | Potential XXE attacks from loading lxml in pandas
[numpy-in-pytorch-modules](python/numpy-in-pytorch-modules.yml) | Python | Uses NumPy functions inside PyTorch modules 
[numpy-in-torch-datasets](python/numpy-in-torch-datasets.yml) | Python | Calls to the Numpy RNG inside of a Torch dataset
[pickles-in-numpy](python/pickles-in-numpy.yml) | Python | Potential arbitrary code execution from NumPy functions reliant on pickling
[pickles-in-pandas](python/pickles-in-pandas.yml) | Python | Potential arbitrary code execution from Pandas functions reliant on pickling
[pickles-in-pytorch](python/pickles-in-pytorch.yml) | Python | Potential arbitrary code execution from PyTorch functions reliant on pickling
[pickles-in-torch-distributed](python/pickles-in-torch-distributed.yml) | Python | Potential arbitrary code execution from PyTorch Distributed functions reliant on pickling
[torch-package](python/torch-package.yml) | Python | Potential arbitrary code execution from torch.package 
[torch-tensor](python/torch-tensor.yml) | Python | Possible parsing issues and inefficiency from improper tensor creation
[waiting-with-torch-distributed](python/waiting-with-torch-distributed.yml) | Python | Possible undefined behavior when not waiting for requests

### CODE REVIEW

File: python/automatic-memory-pinning.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+rules:
+- id: automatic-memory-pinning
+  patterns:
+    - pattern: torch.utils.data.DataLoader(...)
+    - pattern-not-inside: torch.utils.data.DataLoader(..., pin_memory=True, ...)
```
Comment (disconnect3d) [inline python/automatic-memory-pinning.yml:5]: 1) will it work with `x = torch.utils.data; x.DataLoader(...)`?
2) I guess it will miss cases with `kwargs = {'pin_memory': True}; torch.utils.data.DataLoader(**kwargs)`
3) I guess it will miss cases with `torch.utils.data.DataLoader(pin_memory=1)` - does this work in torch?

Is there any way we can catch 2) and 3)? If not, we should document it.

File: python/lxml-in-pandas.yml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+rules:
+- id: lxml-in-pandas
+  patterns:
+    - pattern: pandas.read_html(...)
+    - pattern-not: pandas.read_html(..., flavor="bs4", ...)
+    - pattern-not: pandas.read_html(..., flavor="html5lib", ...)
```
Comment (disconnect3d) [inline python/lxml-in-pandas.yml:6]: What about `**kwargs`, will it be caught? If not, document it pls

File: python/numpy-in-pytorch-modules.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+rules:
+- id: numpy-in-pytorch-modules
+  patterns:
+    - pattern: $RESULT = numpy.$FUNCTION(...)
+    - pattern-inside: |
+        class $MODULE(torch.nn.Module):
```
Comment (disconnect3d) [inline python/numpy-in-pytorch-modules.yml:6]: Will it work if `Module` is imported directly or if it is aliased to sth? We need testcases for this

File: python/numpy-in-torch-datasets.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+rules:
+- id: numpy-in-torch-datasets
+  patterns:
+    - pattern-either:
+      - pattern: |
+          class $X(Dataset):
+            def __getitem__(...):
```
Comment (disconnect3d) [inline python/numpy-in-torch-datasets.yml:7]: It seems we only match `def __getitem__` if its first member of `class $X`, shouldn't there be `...` after `class $X` line?

Pls add testcases for those

File: python/pickles-in-numpy.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
+rules:
+- id: pickles-in-numpy
+  pattern: numpy.load(..., allow_pickle=True)
```
Comment (disconnect3d) [inline python/pickles-in-numpy.yml:3]: 1) Handle `allow_pickle=1` or `allow_pickle=<non zero int>` if they work
2) Handle or document `**kwargs`

File: python/pickles-in-pandas.yml [hunk 6]
```diff
@@ -0,0 +1,11 @@
+rules:
+- id: pickles-in-pandas
+  patterns:
+    - pattern-either:
+      - pattern: pandas.read_pickle(...)
+      - pattern: $DF.to_pickle(...)
```
Comment (disconnect3d) [inline python/pickles-in-pandas.yml:6]: Add pattern: `x = $DF.to_pickle; x(...)`

Comment (disconnect3d) [reply python/pickles-in-pandas.yml:6]: Also add testcase for:
```python
def foo(..., $X=$DF.to_pickle, ...):
    ...
    $X(...)
    ...
```

**Comment by suhacker1** [comment]
Quick update: I wrote an action to run tests with Semgrep as described [here](https://semgrep.dev/docs/writing-rules/testing-rules/). The official Semgrep ruleset from r2c has its own action. I tried implementing a similar [workflow](https://github.com/marketplace/actions/semgrep-rules-test), but ran into [some issues](https://github.com/returntocorp/semgrep-rules-test-action/issues) and decided to opt for this simpler workflow inspired by another Semgrep ruleset [repository](https://github.com/ajinabraham/njsscan/blob/master/.github/workflows/python_test.yml). Feel free to let me know if I should make any major changes to the workflow. I'm still adding tests and looking at edge cases for some of these rules.

File: python/automatic-memory-pinning.yml (hunk 1 above)
Comment (suhacker1) [reply python/automatic-memory-pinning.yml:5]: I updated the rule; it now works with all 3 of those cases.

File: python/lxml-in-pandas.yml (hunk 2 above)
Comment (suhacker1) [reply python/lxml-in-pandas.yml:6]: It catches it.

File: python/numpy-in-pytorch-modules.yml (hunk 3 above)
Comment (suhacker1) [reply python/numpy-in-pytorch-modules.yml:6]: It works! I'll add a test.

File: python/numpy-in-torch-datasets.yml (hunk 4 above)
Comment (suhacker1) [reply python/numpy-in-torch-datasets.yml:8]: Oddly enough, it seems to work without it, but I went ahead and made the change in case. I'll go ahead and add a test case for different permutations of methods.

File: python/pickles-in-numpy.yml (hunk 5 above)
Comment (suhacker1) [reply python/pickles-in-numpy.yml]: The rule has been updated to handle 1. It catches case 2.

File: python/pickles-in-pandas.yml (hunk 6 above)
Comment (suhacker1) [reply python/pickles-in-pandas.yml:6]: I added two new patterns as that test case wouldn't have been caught by the original rule.

File: python/automatic-memory-pinning.yml (hunk 1 above)
Comment (suhacker1) [reply python/automatic-memory-pinning.yml]: It also now works with any non-zero integer.

Comment (suhacker1) [reply python/automatic-memory-pinning.yml]: I wasn't able to capture all the cases with the original rule. More concretely, I wasn't able to write a rule that would capture instances when `pin_memory` isn't set at all and when it is set to `False`. As I was doing more research into this issue, it turns out there are a lot of cases where automatic memory pinning is not possible, so I decided to flag cases where the argument isn't set. That way, the rule will find instances where the implications of automatically pinning memory were not considered. Let me know if you think I should revert this rule.

**Comment by suhacker1** [comment]
All the rules I added now have tests and have been updated to reflect the edge cases. Let me know if there's anything else I should add. Thanks!

**Review by disconnect3d — APPROVED** [review]
